biscuit-auth/src/crypto/mod.rs

```diff
@@ -176,55 +247,68 @@ impl PublicKey {
     }
 }
 
-impl PartialEq for PublicKey {
-    fn eq(&self, other: &Self) -> bool {
-        self.0.to_bytes() == other.0.to_bytes()
+#[derive(Clone, Debug)]
+pub struct Signature(pub(crate) Vec<u8>);
+
+impl Signature {
+    pub fn from_bytes(data: &[u8]) -> Result<Self, error::Format> {
+        Ok(Signature(data.to_owned()))
+    }
+
+    pub(crate) fn from_vec(data: Vec<u8>) -> Self {
+        Signature(data)
     }
-}
 
-impl Hash for PublicKey {
-    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
-        (crate::format::schema::public_key::Algorithm::Ed25519 as i32).hash(state);
-        self.0.to_bytes().hash(state);
+    pub fn to_bytes(&self) -> &[u8] {
+        &self.0[..]
     }
 }
 
 impl FromStr for PublicKey {
     type Err = error::Token;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let (_, bytes) = biscuit_parser::parser::public_key(s)
+        let (_, public_key) = biscuit_parser::parser::public_key(s)
             .finish()
             .map_err(biscuit_parser::error::LanguageError::from)?;
-        Ok(PublicKey::from_bytes(&bytes)?)
+        Ok(PublicKey::from_bytes(
+            &public_key.key,
+            match public_key.algorithm {
+                biscuit_parser::builder::Algorithm::Ed25519 => Algorithm::Ed25519,
+                biscuit_parser::builder::Algorithm::Secp256r1 => Algorithm::Secp256r1,
+            },
+        )?)
     }
 }
 
 impl Display for PublicKey {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "ed25519/{}", hex::encode(self.to_bytes()))
+        match self {
+            PublicKey::Ed25519(key) => write!(f, "ed25519/{}", hex::encode(key.to_bytes())),
```

Comment:
i think there should be a dedicated function for datalog rendering. Using `Display` here to generate datalog code is a bit fragile.
Also, both public key formats have a `print()` method that could be used to avoid duplication (if used for datalog formatting, it should also be named accordingly

Reply:
fixed in ae3d779